src/Planner/PlannerAggregation.cpp

```diff
@@ -183,6 +185,87 @@ void resolveGroupingFunctions(QueryTreeNodePtr & query_node,
     resolveGroupingFunctions(query_node, group_by_kind, aggregation_keys, grouping_sets_parameters_list, planner_context);
 }
 
+}
+
+void resolveGroupingFunctions(QueryTreeNodePtr & query_node, const PlannerContext & planner_context)
+{
+    auto & query_node_typed = query_node->as<QueryNode &>();
+
+    std::unordered_set<std::string> used_aggregation_keys;
+    Names aggregation_keys;
+    GroupingSetsParamsList grouping_sets_parameters_list;
+
+    QueryTreeNodeToName group_by_node_to_name;
+
+    /// Add expressions from GROUP BY
+
+    if (query_node_typed.hasGroupBy())
+    {
+        if (query_node_typed.isGroupByWithGroupingSets())
+        {
+            for (const auto & grouping_set_keys_list_node : query_node_typed.getGroupBy().getNodes())
+            {
+                auto & grouping_set_keys_list_node_typed = grouping_set_keys_list_node->as<ListNode &>();
+                grouping_sets_parameters_list.emplace_back();
+                auto & grouping_sets_parameters = grouping_sets_parameters_list.back();
+
+                for (auto & grouping_set_key_node : grouping_set_keys_list_node_typed.getNodes())
+                {
+                    auto grouping_set_key_name = calculateActionNodeName(grouping_set_key_node, planner_context, group_by_node_to_name);
+                    grouping_sets_parameters.used_keys.push_back(grouping_set_key_name);
+
+                    if (used_aggregation_keys.contains(grouping_set_key_name))
+                        continue;
+
+                    aggregation_keys.push_back(grouping_set_key_name);
+                    used_aggregation_keys.insert(aggregation_keys.back());
+                }
+            }
+
+            for (auto & grouping_sets_parameter : grouping_sets_parameters_list)
+            {
+                NameSet grouping_sets_used_keys;
```

Comment:
Why do we use names? Maybe it's better to do it via query tree nodes in the new analyzer?

Reply:
Updated.